packages/hash/api/src/graph/knowledge/primitive/entity.ts

```diff
@@ -0,0 +1,643 @@
+    params;
+
+  const entitiesInTree = linkedTreeFlatten<
+    EntityDefinition,
+    LinkedEntityDefinition,
+    "linkedEntities",
+    "entity"
+  >(
+    {
+      entityTypeId,
+      entityProperties: properties,
+      linkedEntities,
+    },
+    "linkedEntities",
+    "entity",
+  );
+
+  /**
+   * @todo Once the graph API validates the required links of entities on creation, this may have to be reworked in order
+   *   to create valid entities.
+   *   this code currently creates entities first, then links them together.
+   *   See https://app.asana.com/0/1202805690238892/1203046447168478/f
+   */
+  const entities = await Promise.all(
+    entitiesInTree.map(async (definition) => ({
+      link: definition.meta
+        ? {
+            parentIndex: definition.parentIndex,
+            meta: definition.meta,
+          }
+        : undefined,
+      entity: await getOrCreateEntity(
+        { graphApi },
+        {
+          ownedById,
+          entityDefinition: definition,
+          actorId,
+        },
+      ),
+    })),
+  );
+
+  let rootEntity: Entity;
+  if (entities[0]) {
+    // First element will be the root entity.
+    rootEntity = entities[0].entity;
+  } else {
+    throw new ApolloError(
+      "Could not create entity tree",
+      "INTERNAL_SERVER_ERROR",
+    );
+  }
+
+  await Promise.all(
+    entities.map(async ({ link, entity }) => {
+      if (link) {
+        const parentEntity = entities[link.parentIndex];
+        if (!parentEntity) {
+          throw new ApolloError("Could not find parent entity");
+        }
+        const linkEntityType = await getEntityTypeById(
+          { graphApi },
+          {
+            entityTypeId: link.meta.linkEntityTypeId,
+          },
+        );
+
+        // links are created as an outgoing link from the parent entity to the children.
+        await createLinkEntity(
+          { graphApi },
+          {
+            linkEntityType,
+            leftEntityId: parentEntity.entity.metadata.editionId.baseId,
+            rightEntityId: entity.metadata.editionId.baseId,
+            leftToRightOrder: link.meta.index ?? undefined,
+            ownedById,
+            actorId,
+          },
+        );
+      }
+    }),
+  );
+
+  return rootEntity;
+};
+
+/**
+ * Update an entity.
+ *
+ * @param params.properties - the properties object of the entity
+ * @param params.actorId - the id of the account that is updating the entity
+ */
+export const updateEntity: ImpureGraphFunction<
+  {
+    entity: Entity;
+    properties: PropertyObject;
+    actorId: AccountId;
+  },
+  Promise<Entity>
+> = async ({ graphApi }, params) => {
+  const { entity, properties, actorId } = params;
+
+  const { data: metadata } = await graphApi.updateEntity({
+    actorId,
+    entityId: entity.metadata.editionId.baseId,
+    /** @todo: this shouldn't be required */
+    entityTypeId: entity.metadata.entityTypeId,
+    properties,
+  });
```

Comment:
The `archived` field will be required, could you add this here as well, please?

Reply:
Sure, I've addressed this in a squashed commit.